community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/spi/RepeatableReadQueryContext.scala

```diff
@@ -113,11 +114,51 @@ final class RepeatableReadQueryContext(inner: QueryContext, locker: Locker) exte
     locker.acquireLock(relationshipOps.getByInnerId(id))
   }
 
-  private def lockAll[T <: PropertyContainer](iter: Iterator[T]): Iterator[T] = iter.map {
-    item =>
-      locker.acquireLock(item)
-      item
+
+  private def lockAllNodesAndFilter(inner: Iterator[Node])(f: Node => Boolean): Iterator[Node] = new Iterator[Node] {
+    var nextNode: Option[Node] = null
```

Comment:
It seems to me that if we have an `Option`, we use `None` for whatever `null` would mean for pretty much any other object, and that this distinction is part of why `Option` is such a nice thing. Why would we violate that? Is it because we need to support three different values for interpretation? Wouldn't an enum work better for that situation?